bazel/generate_objc.bzl

```diff
@@ -0,0 +1,243 @@
+    dir_out = _join_directories([
+        str(ctx.genfiles_dir.path + proto_root), label_package, _GENERATED_PROTOS_DIR
+    ])
+
+    arguments = []
+    if ctx.executable.plugin:
+        arguments += get_plugin_args(
+            ctx.executable.plugin,
+            [],
+            dir_out,
+            False,
+        )
+        tools = [ctx.executable.plugin]
+    arguments += ["--objc_out=" + dir_out]
+    arguments += get_include_protoc_args(protos)
+
+    # Include the output directory so that protoc puts the generated code in the
+    # right directory.
+    arguments += ["--proto_path={0}{1}".format(dir_out, proto_root)]
+    arguments += ["--proto_path={}".format(_get_directory_from_proto(proto)) for proto in protos]
+    arguments += [_get_srcs_file_path(proto) for proto in protos]
+
+    # create a list of well known proto files if the argument is non-None
+    well_known_proto_files = []
+    if ctx.attr.use_well_known_protos:
+        f = ctx.attr.well_known_protos.files.to_list()[0].dirname
+        arguments += ["-I{0}".format(f + "/../..")]
+        well_known_proto_files = [
+            f
+            for f in ctx.attr.well_known_protos.files.to_list()
+        ]
+    ctx.actions.run(
+        inputs = protos + well_known_proto_files,
+        tools = tools,
+        outputs = out_files,
+        executable = ctx.executable._protoc,
+        arguments = arguments,
+    )
+
+    return struct(files = depset(out_files))
+
+def _get_file_name_from_proto(proto):
+    return proto.path.rpartition("/")[2]
+
+def _get_slash_from_proto(proto):
+    return proto.path.rpartition("/")[1]
+
+def _get_directory_from_proto(proto):
+    return proto.path.rpartition("/")[0]
+
+def _strip_package_and_to_camel_case(label_package, file):
+    return to_upper_camel_with_extension(
+        _strip_package_from_path(label_package, file),
+        "proto"
+    )
+
+def _strip_package_from_path(label_package, file):
+    prefix_len = 0
+    if not file.is_source and file.path.startswith(file.root.path):
+        prefix_len = len(file.root.path) + 1
+
+    path = file.path
+    file_name = ""
+    if len(label_package) == 0:
+        file_name = path
+    if not path.startswith(label_package + "/", prefix_len) and len(label_package) > 0:
+        return "//" + file.path
+    if len(label_package) > 0:
+        file_name = path[prefix_len + len(label_package + "/"):]
+    
+    return file_name
+
+def _get_srcs_file_path(file):
+    if not file.is_source and file.path.startswith(file.root.path):
+        return file.path[len(file.root.path) + 1:]
+    return file.path
+
+def _join_directories(directories):
+    massaged_directories = [directory for directory in directories if len(directory) != 0]
+    return "/".join(massaged_directories)
+
+
+generate_objc = rule(
+    attrs = {
+        "srcs": attr.label_list(
+            mandatory = True,
+            allow_empty = False,
+            providers = [ProtoInfo],
```

Comment:
Is `ProtoInfo` a build-in provider type?

Reply:
@muxi Yes